### PR TITLE
Convert storm runtime to base to allow fini and implement storm websocket that requires cleanup

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/stormtypes.py [hunk 1]
```diff
@@ -963,7 +963,6 @@ def getObjLocals(self):
             'trycast': self.trycast,
         }
 
-    @stormfunc(readonly=True)
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:966]: Why is this change being made?

File: synapse/lib/stormhttp.py [hunk 2]
```diff
@@ -10,9 +10,74 @@
 import synapse.exc as s_exc
 import synapse.common as s_common
 
+import synapse.lib.base as s_base
 import synapse.lib.msgpack as s_msgpack
 import synapse.lib.stormtypes as s_stormtypes
 
+class WebSocket(s_base.Base, s_stormtypes.StormType):
+
```
Comment (vEpiphyte) [inline synapse/lib/stormhttp.py:18]: Class docstring

File: synapse/lib/stormhttp.py [hunk 3]
```diff
@@ -10,9 +10,74 @@
 import synapse.exc as s_exc
 import synapse.common as s_common
 
+import synapse.lib.base as s_base
 import synapse.lib.msgpack as s_msgpack
 import synapse.lib.stormtypes as s_stormtypes
 
+class WebSocket(s_base.Base, s_stormtypes.StormType):
```
Comment (vEpiphyte) [inline synapse/lib/stormhttp.py:17]: @s_stormtypes.registry.registerType

File: synapse/lib/stormhttp.py [hunk 4]
```diff
@@ -10,9 +10,74 @@
...
+            _type, data, extra = await self.resp.receive()
+            if _type == aiohttp.WSMsgType.BINARY:
+                return (True, json.loads(data))
+            if _type == aiohttp.WSMsgType.TEXT:
+                return (True, json.loads(data.encode()))
+            if _type == aiohttp.WSMsgType.CLOSED: # pragma: no cover
+                return (True, None)
+            # pragma: no cover
+            mesg = f'WebSocket RX unhandled type: {_type.name}'
+            return (False, ('BadMesgFormat', {'mesg': mesg}))
```
Comment (vEpiphyte) [inline synapse/lib/stormhttp.py:72]:
```
return (False, ('BadMesgFormat', {'mesg': return (False, ('BadMesgFormat', {'mesg': mesg}))}))  # pragma: no cover
```

**Review by vEpiphyte — CHANGES_REQUESTED** [review]
few doc changes and error line that can be collapsed down

File: synapse/lib/stormhttp.py [hunk 5]
```diff
@@ -10,9 +10,75 @@
...
+            await self.resp.send_bytes(json.dumps(mesg).encode())
+            return (True, None)
+
+        except asyncio.CancelledError: # pragma: no cover
+            raise
+
+        except Exception as e: # pragma: no cover
+            return s_common.retnexc(e)
+
+    async def rx(self, timeout=None):
```
Comment (Cisphyx) [inline synapse/lib/stormhttp.py:63]: timeout is unused

File: synapse/lib/stormhttp.py [hunk 6]
```diff
@@ -10,9 +10,75 @@
...
+
+        {'name': 'tx', 'desc': 'Transmit a message over the web socket.',
+         'type': {'type': 'function', '_funcname': 'tx',
+                  'args': (
+                      {'name': 'mesg', 'type': 'dict', 'desc': 'A JSON compatible message.', },
+                  ),
+                  'returns': {'type': 'list', 'desc': 'An ($ok, $valu) tuple.'}}},
+
+        {'name': 'rx', 'desc': 'Receive a message from the web socket.',
+         'type': {'type': 'function', '_funcname': 'tx',
```
Comment (Cisphyx) [inline synapse/lib/stormhttp.py:34]: funcname should be `rx`

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**